### PR TITLE
[METRICS] Support expose Prometheus metrics

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Support expose Prometheus metrics

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/dependencyList [hunk 1]
```diff
@@ -30,8 +30,16 @@ htrace-core4/4.1.0-incubating//htrace-core4-4.1.0-incubating.jar
 jackson-annotations/2.11.4//jackson-annotations-2.11.4.jar
 jackson-core/2.11.4//jackson-core-2.11.4.jar
 jackson-databind/2.11.4//jackson-databind-2.11.4.jar
+jakarta.servlet-api/4.0.4//jakarta.servlet-api-4.0.4.jar
 javassist/3.18.1-GA//javassist-3.18.1-GA.jar
 jcl-over-slf4j/1.7.30//jcl-over-slf4j-1.7.30.jar
+jetty-http/9.4.40.v20210413//jetty-http-9.4.40.v20210413.jar
```
Comment (yaooqinn) [inline dev/dependencyList:36]: why not jetty 11?

Comment (pan3793) [reply dev/dependencyList:36]: Jetty 11 required JVM 11+, see https://www.eclipse.org/jetty/download.php

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 2]
```diff
@@ -41,11 +44,11 @@ class MetricsSystemSuite extends KyuubiFunSuite {
     }
   }
 
-  test("MetricsSystem") {
+  test("metrics - JsonReporter") {
     val reportPath = Utils.createTempDir()
     val conf = KyuubiConf()
       .set(MetricsConf.METRICS_ENABLED, true)
-      .set(MetricsConf.METRICS_REPORTERS, ReporterType.values.map(_.toString).toSeq)
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:48]: can we still have all non-PrometheusReporters?

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 3]
```diff
@@ -63,4 +66,30 @@ class MetricsSystemSuite extends KyuubiFunSuite {
...
+      .set(MetricsConf.METRICS_ENABLED, true)
+      .set(MetricsConf.METRICS_REPORTERS, Seq("PROMETHEUS"))
+      .set(MetricsConf.METRICS_PROMETHEUS_PORT, testPort)
+      .set(MetricsConf.METRICS_PROMETHEUS_PATH, testContextPath)
+    val metricsSystem = new MetricsSystem()
+    metricsSystem.initialize(conf)
+    metricsSystem.start()
+    metricsSystem.incAndGetCount(MetricsConstants.STATEMENT_TOTAL)
+
+    metricsSystem.decAndGetCount(MetricsConstants.STATEMENT_TOTAL)
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:83]: can we remove these?

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 4]
```diff
@@ -63,4 +66,30 @@ class MetricsSystemSuite extends KyuubiFunSuite {
     metricsSystem.stop()
   }
 
+  test("metrics - PrometheusReporter") {
+    val testPort = 10019
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:70]: can we use random port for test cases

**Comment by yaooqinn** [comment]
where is the codecov report? 🤓

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:70]: make sense

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:83]: ok, will update

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala (hunk 2 above)
Comment (pan3793) [reply kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:48]: update to "metrics - PrometheusReporter" and "metrics - other reporters"

File: kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    httpServer.setHandler(context)
+
+    new DropwizardExports(registry).register(bridgeRegistry)
+    val metricsServlet = new MetricsServlet(bridgeRegistry)
+    context.addServlet(new ServletHolder(metricsServlet), contextPath)
+
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = {
```
Comment (yaooqinn) [inline kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala:55]: better syncronized here and `stop`

File: docs/deployment/settings.md [hunk 6]
```diff
@@ -194,7 +194,9 @@ kyuubi\.metrics<br>\.console\.interval|<div style='width: 65pt;word-wrap: break-
...
-kyuubi\.metrics<br>\.reporters|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>JSON</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>A comma separated list for all metrics reporters<ul> <li>CONSOLE - ConsoleReporter which outputs measurements to CONSOLE periodically.</li> <li>JMX - JmxReporter which listens for new metrics and exposes them as MBeans.</li>  <li>JSON - JsonReporter which outputs measurements to json file periodically.</li> <li>SLF4J - Slf4jReporter which outputs measurements to system log periodically.</li></ul></div>|<div style='width: 30pt'>seq</div>|<div style='width: 20pt'>1.2.0</div>
+kyuubi\.metrics<br>\.prometheus\.path|<div style='width: 65pt;word-wrap: break-word;white-space: normal'>/metrics</div>|<div style='width: 170pt;word-wrap: break-word;white-space: normal'>URI context path of prometheus metrics HTTP server</div>|<div style='width: 30pt'>string</div>|<div style='width: 20pt'>1.2.0</div>
```
Comment (yaooqinn) [inline docs/deployment/settings.md:197]: cp these lines to metrics.md

File: kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    val metricsServlet = new MetricsServlet(bridgeRegistry)
+    context.addServlet(new ServletHolder(metricsServlet), contextPath)
+
+    super.initialize(conf)
+  }
+
+  override def start(): Unit = synchronized {
+    if (!isStarted) {
+      try {
+        httpServer.start()
```
Comment (yaooqinn) [inline kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala:59]: add a info log?

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 8]
```diff
@@ -41,26 +44,55 @@ class MetricsSystemSuite extends KyuubiFunSuite {
...
+      .set(MetricsConf.METRICS_ENABLED, true)
+      .set(MetricsConf.METRICS_REPORTERS, Seq(ReporterType.PROMETHEUS.toString))
+      .set(MetricsConf.METRICS_PROMETHEUS_PORT, 0) // random port
+      .set(MetricsConf.METRICS_PROMETHEUS_PATH, testContextPath)
+    val metricsSystem = new MetricsSystem()
+    metricsSystem.initialize(conf)
+    metricsSystem.start()
+
+    val testPort = metricsSystem.getServices
+      .find(_.isInstanceOf[PrometheusReporterService]).get
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:60]: can simply use head here for a test case

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 9]
```diff
@@ -41,26 +44,55 @@ class MetricsSystemSuite extends KyuubiFunSuite {
...
+      .asInstanceOf[PrometheusReporterService]
+      .httpServer.getURI.getPort
+
+    metricsSystem.registerGauge(MetricsConstants.CONN_OPEN, 2021, 0)
+
+    val client: HttpClient = new HttpClient
+    client.start()
+    val res: ContentResponse = client.GET(s"http://localhost:$testPort$testContextPath")
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:68]: replace this with` ...httpServer.getURI$...`

**Comment by yaooqinn** [comment]
LGTM, only minor revisions.

File: docs/deployment/settings.md (hunk 6 above)
Comment (pan3793) [reply docs/deployment/settings.md:197]: done

File: kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala (hunk 7 above)
Comment (pan3793) [reply kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala:59]: added

**Comment by pan3793** [comment]
@yaooqinn Thanks for reviewing, the last commit has addressed all comments, and good night :)

File: kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala [hunk 10]
```diff
@@ -56,16 +56,15 @@ class MetricsSystemSuite extends KyuubiFunSuite {
...
 
-    val testPort = metricsSystem.getServices
-      .find(_.isInstanceOf[PrometheusReporterService]).get
-      .asInstanceOf[PrometheusReporterService]
-      .httpServer.getURI.getPort
-
     metricsSystem.registerGauge(MetricsConstants.CONN_OPEN, 2021, 0)
 
+    val prometheusHttpServer = metricsSystem.getServices
+      .find(_.isInstanceOf[PrometheusReporterService]).head
```
Comment (yaooqinn) [inline kyuubi-metrics/src/test/scala/org/apache/kyuubi/metrics/MetricsSystemSuite.scala:62]: I MEAN getServices.head

File: kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        info(s"Prometheus metrics HTTP server has started at port ${httpServer.getURI.getPort}.")
```
Comment (yaooqinn) [inline kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/PrometheusReporterService.scala:60]: simple print uri?

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.2.0
